extensions/standard-processors/processors/RetryFlowFile.cpp

```diff
@@ -0,0 +1,212 @@
+    FailOnNonNumericalOverwrite,
+    ReuseMode,
+  });
+  setSupportedRelationships({
+    Retry,
+    RetriesExceeded,
+    Failure,
+  });
+}
+
+void RetryFlowFile::onSchedule(core::ProcessContext* context, core::ProcessSessionFactory* /* sessionFactory */) {
+  context->getProperty(RetryAttribute.getName(), retry_attribute_);
+  context->getProperty(MaximumRetries.getName(), maximum_retries_);
+  context->getProperty(PenalizeRetries.getName(), penalize_retries_);
+  context->getProperty(FailOnNonNumericalOverwrite.getName(), fail_on_non_numerical_overwrite_);
+  context->getProperty(ReuseMode.getName(), reuse_mode_);
+  readDynamicPropertyKeys(context);
+}
+
+void RetryFlowFile::onTrigger(core::ProcessContext* context, core::ProcessSession* session) {
+  std::shared_ptr<FlowFileRecord> flow_file = std::static_pointer_cast<FlowFileRecord> (session->get());
+  if (!flow_file) {
+    return;
+  }
+
+  bool failure_due_to_non_numerical_retry;
+  uint64_t retry_property_value;
+  std::tie(retry_property_value, failure_due_to_non_numerical_retry) = getRetryPropertyValue(flow_file);
+  if (failure_due_to_non_numerical_retry) {
+    session->transfer(flow_file, Failure);
+    return;
+  }
+  if (updateUUIDMarkerAndCheckFailOnReuse(flow_file)) {
+    session->transfer(flow_file, Failure);
+    return;
+  }
+
+  if (retry_property_value < maximum_retries_) {
+    try {
+      flow_file->setAttribute(retry_attribute_, std::to_string(gsl::narrow_cast<uint64_t>(retry_property_value + 1)));
+    }
+    catch(const gsl::narrowing_error& e) {
+      logger_->log_error("Narrowing Exception: %s", e.what());
+      session->transfer(flow_file, Failure);
+      return;
+    }
+    if (penalize_retries_) {
+      session->penalize(flow_file);
+    }
+    session->transfer(flow_file, Retry);
+    return;
+  }
+  if (!setRetriesExceededAttributesOnFlowFile(context, flow_file)) {
+    session->transfer(flow_file, Failure);
+    yield();
+    return;
+  }
+  session->transfer(flow_file, RetriesExceeded);
+}
+
+void RetryFlowFile::readDynamicPropertyKeys(core::ProcessContext* context) {
+  exceeded_flowfile_attribute_keys.clear();
+  const std::vector<std::string> dynamic_prop_keys = context->getDynamicPropertyKeys();
+  logger_->log_info("RetryFlowFile registering %d keys", dynamic_prop_keys.size());
+  for (const auto& key : dynamic_prop_keys) {
+    exceeded_flowfile_attribute_keys.emplace_back(core::PropertyBuilder::createProperty(key)->withDescription("auto generated")->supportsExpressionLanguage(true)->build());
+    logger_->log_info("RetryFlowFile registered attribute '%s'", key);
+  }
+}
+
+// Returns (1, true) on non-numerical or out-of-bounds retry value
+std::pair<uint64_t, bool> RetryFlowFile::getRetryPropertyValue(const std::shared_ptr<FlowFileRecord>& flow_file) {
+  std::string value_as_string;
+  try {
+    if (flow_file->getAttribute(retry_attribute_, value_as_string)) {
+      return std::make_pair(std::stoul(value_as_string), false);
```

Comment:
watch our as `std::stoul` does many counterintuitive things, e.g. it doesn't try to parse till the end of the string, meaning `std::stoul("1banana")` will happily return `1`, moreover `std::stoul("-1")` happily returns the max `unsigned long`, check out #797 `ValueParser.h` which aims to tackle these problems

Reply:
Well, this is not merged yet, so using `std::stoull` is the best I can do. I do not think having the issues you mentioned should be a big problem though, I can live with those conversions.

Reply:
Updated using `ValueParser`.